admiral/envs/components/observer.py

```diff
@@ -1,41 +1,125 @@
 
+from abc import ABC, abstractmethod, abstractproperty
+
 from gym.spaces import Box, Discrete, Dict
 import numpy as np
 
 from admiral.envs.components.agent import HealthObservingAgent, LifeObservingAgent, \
     AgentObservingAgent, PositionObservingAgent, SpeedAngleObservingAgent, VelocityObservingAgent, \
     ResourceObservingAgent, TeamObservingAgent, BroadcastObservingAgent, SpeedAngleAgent, \
-    VelocityAgent, BroadcastingAgent
-
-# --------------------- #
-# --- Communication --- #
-# --------------------- #
+    VelocityAgent, BroadcastingAgent, Agent
 
-class BroadcastObserver:
+class Observer(ABC):
     """
-    Observe the broadcast state of broadcasting agents.
+    Base observer class provides the interface required of all observers. Setup
+    the agents' observation space according to the Observer's channel.
+
+        agents (dict):
+            The dictionary of agents.
     """
     def __init__(self, agents=None, **kwargs):
         self.agents = agents
+    
+    # TODO: WE hsould probably make two types of observers, one for simple space
```

Comment:
Make two observers, one for simple channel space where the space does not depend on the type of other (Life, Position, and Team) and another for more complicated channel space where the space does depend on the type of the other.

Reply:
Remove this todo